### DISCUSSION TITLE
EHCI/OHCI on allwinner t113

### DISCUSSION THREAD

**Discussion by robots** [hathach/tinyusb#1599] [Q&A]
Hi,
I am working on baremetal allwinner t113. The usb controller is ohci/ehci pair. 

There are few questions i have:
- how much NXP specific is ehci driver ? I have noticed few things:
1.  hcd_port_speed_get is nxp specific function
2.  TUP_USBIP_CHIPIDEA_HS needs to be defined (otherwise the linked lists are wrongly initialized)
3. missing  regs->config_flag = 1; after initialization is complete.
- there is one place "bool hcd_setup_send(uint8_t rhport, uint8_t dev_addr, uint8_t const setup_packet[8])", where data is not located in CFG_TUSB_MEM_SECTION defined. (CFG_TUSB_MEM_SECTION is non cachable in my system)
- Is is possible to use "just" EHCI and ignore the OHCI part ? Or it is mandatory to switch to OHCI once low/full speed device is detected?

Right now i have been able to get to point, that I get interrupt that device is attached, but any transfer fails. I assume that phy is working, ehci is enabled correctly. Device i am connecting is Keyboard or usb flash drive - both behave similarly.

```
regs->portsc = 1403
  Current Connect Status: 1
  Connect Status Change : 1
  Port Enabled          : 0
  Port Enabled Change   : 0
  Port Reset            : 0
  Port Power            : 1
[0:] USBH DEVICE ATTACH
Low Speed    <---- manually set in "hcd_port_speed_get"
[0:0] Open EP0 with Size = 8
Get 8 byte of Device Descriptor
[0:0] Get Descriptor: 80 06 00 01 00 00 08 00
on EP 00 with 0 bytes
[0:0] Control FAILED
on EP 00 with 32768 bytes
[0:0] Control FAILED
```

**Comment by hathach** [comment]
> Hi, I am working on baremetal allwinner t113. The usb controller is ohci/ehci pair.

This is very advanced CPU, I am glad that TinyUSB is getting into these. 

> There are few questions i have:
> 
> * how much NXP specific is ehci driver ? I have noticed few things:
> 
> 1. hcd_port_speed_get is nxp specific function
> 2. TUP_USBIP_CHIPIDEA_HS needs to be defined (otherwise the linked lists are wrongly initialized)
> 3. missing  regs->config_flag = 1; after initialization is complete.

Indeed current EHCI has some NXP specific code, since it is the only EHCI chip I worked with. And yes, your 1 and 2 are spot-on. For
1. NXP has embedded TT which allow EHCI to work with full/low speed thus it has additional bit for those speed. It should be wrapped up with `#ifdef TUP_USBIP_CHIPIDEA_HS`, feel free to make the change as necessary to adopt allwinner
2. Right, EHCI smallest list is 256 entries, which is not really necessary, I tried to minimize the SRAM usage and use smallest possible since NXP MCU has limited SRAM. And since it is the only EHCI mcu I work so far, I haven't bothered to test with 256 list, again feel free to add #ifdef and make necessary changes as it fit and make an PR. I will be happy to merge this
3. feel free to add it if it is part of the specs, I may overlook it since it is working with NXP chip

> * there is one place "bool hcd_setup_send(uint8_t rhport, uint8_t dev_addr, uint8_t const setup_packet[8])", where data is not located in CFG_TUSB_MEM_SECTION defined. (CFG_TUSB_MEM_SECTION is non cachable in my system)

Feel free to make changes as needed. Actual we could make these into cache memory for better performance as well and add clean_invalidate() where needed. We did it with the device driver already, ehci/host driver isn't just popular enough (and have time) to do so

https://github.com/hathach/tinyusb/blob/251ce9899039fbdf8e749342bc6504e7438d2e88/src/portable/chipidea/ci_hs/dcd_ci_hs.c#L51-L55

> * Is is possible to use "just" EHCI and ignore the OHCI part ? Or it is mandatory to switch to OHCI once low/full speed device is detected?

Per specs, EHCI can only handle HS traffic and timing. In order to do FS/LS, chipidea integrate an Transaction Translator (TT) into their custom design controller (used by NXP), which serve as HS-to-FS/LS hub. I am not familiar with allwinner, but if it does not have an embedded TT, we probably need to hand over FS/LS to OHCI.

Note: currently tinyusb can only support 1 active host controller due to the fact they use the same naming API. It may cause issue if you want to have both EHCI/OHCI running at the same time. I will try to update the host driver later on to enable multiple controllers. Though this is rather rare with MCU since they pretty much only uses 1 type of controller save space. 

> Right now i have been able to get to point, that I get interrupt that device is attached, but any transfer fails. I assume that phy is working, ehci is enabled correctly. Device i am connecting is Keyboard or usb flash drive - both behave similarly.
> 
> ```
> regs->portsc = 1403
>   Current Connect Status: 1
>   Connect Status Change : 1
>   Port Enabled          : 0
>   Port Enabled Change   : 0
>   Port Reset            : 0
>   Port Power            : 1
> [0:] USBH DEVICE ATTACH
> Low Speed    <---- manually set in "hcd_port_speed_get"
> [0:0] Open EP0 with Size = 8
> Get 8 byte of Device Descriptor
> [0:0] Get Descriptor: 80 06 00 01 00 00 08 00
> on EP 00 with 0 bytes
> [0:0] Control FAILED
> on EP 00 with 32768 bytes
> [0:0] Control FAILED
> ```

That is some progress, you should better test with an high speed device such as MSC. since low speed may require OCHI to work with. The incorrect speed report from hcd_port_speed_get() definitely need some changes to work with std EHCI. Unfortunately, I am not familiar with allwinner and does not have hw/time to set it up properly to test it with you at the moment. But, please keep posting the update, I will try to help whenever I could.

**Reply by robots** [reply]
TinyUSB is one of very few host stacks that i could find, so i gave it a shot. The code is very nice written, very easy to debug, very easy to understand!

I thought that EHCI will be trouble on its own. So I switched to OHCI as more easier thing to get going. I switched to OHCI driver from pull request from @Ryzee119, added few places to clear/invalidate caches and .... works :)

```
@@ -308,10 +308,13 @@ static void gtd_init(ohci_gtd_t* p_td, uint8_t* data_ptr, uint16_t total_bytes)
   p_td->delay_interrupt        = OHCI_INT_ON_COMPLETE_NO;
   p_td->condition_code         = OHCI_CCODE_NOT_ACCESSED;
 
   p_td->current_buffer_pointer = data_ptr;
   p_td->buffer_end             = total_bytes ? (data_ptr + total_bytes-1) : data_ptr;
+
+		extern void L1_CleanInvalidateDCache_by_Addr(volatile void *addr, int32_t dsize);
+	L1_CleanInvalidateDCache_by_Addr(data_ptr, total_bytes);
 }
 
 static ohci_ed_t * ed_from_addr(uint8_t dev_addr, uint8_t ep_addr)
 {
   if ( tu_edpt_number(ep_addr) == 0 ) return &ohci_data.control[dev_addr].ed;
```
One problem is that in the xfer finish interrupt the buffer address is not available anymore. current_buffer_pointer is set to 0. It would be nice to invalidate cache in interrupt handler, that way the higher levels do not need to care anymore. (Save the buffer pointer somewhere in the linked list, take it out in ISR) Right now i invalidate cache in HID layer.


> Note: currently tinyusb can only support 1 active host controller due to the fact they use the same naming API.

This will probably need some multiplexing layer. Check ownership of ports and calling appropriate xfer function.

> Per specs, EHCI can only handle HS traffic and timing. In order to do FS/LS, chipidea integrate an Transaction Translator (TT) into their custom design controller (used by NXP), which serve as HS-to-FS/LS hub. I am not familiar with allwinner, but if it does not have an embedded TT, we probably need to hand over FS/LS to OHCI.

What puzzles me is the official melis os for d1s (t113 is the same chip with different cpu core arm vs riscv). They only compile ehci stuff.
https://github.com/Tina-Linux/d1s-melis/blob/master/ekernel/drivers/hal/source/usb/host/Makefile
But the documentation is very sparse on details and almost nonexistent.

In Melis os they use parts of Linux kernel usb stack, and parts of ehci drivers. I will have to dive into that.

**Comment by Ryzee119** [comment]
@robots are you using a particular dev board? I wouldnt mind grabbing one of these
